### PR TITLE
Separating executor for ReactiveSocket and DuplexConnection

### PR DESCRIPTION
**Pull request by lehecka**
We need to be able to run ReactiveSocket and DuplexConnection each run on a different Executor.
This change introduces dispatching frames (FrameProcessor interface) to ConnectionAutomaton via ExecutorBase. Also Executor to DuplexConnection ctor is now explicit.

### CODE REVIEW

File: src/ConnectionAutomaton.cpp [hunk 1]
```diff
@@ -260,6 +277,15 @@ void ConnectionAutomaton::closeStreams(StreamCompletionSignal signal) {
 }
 
 void ConnectionAutomaton::processFrame(std::unique_ptr<folly::IOBuf> frame) {
+  auto thisPtr = this->shared_from_this();
+  auto movedFrame = folly::makeMoveWrapper(frame);
```
Comment (yschimke) [inline src/ConnectionAutomaton.cpp]: this is making work for us to revert, I'll try to see if I can fix travis for 4.9

Comment (yschimke) [inline src/ConnectionAutomaton.cpp]: Don't let this block you, its a current problem, not a new problem

File: src/StandardReactiveSocket.cpp [hunk 2]
```diff
@@ -195,6 +211,23 @@ void StandardReactiveSocket::createResponder(
     ConnectionAutomaton& connection,
     StreamId streamId,
     std::unique_ptr<folly::IOBuf> serializedFrame) {
+  debugCheckCorrectExecutor();
+
+  if (streamId != 0) {
+    if (nextStreamId_ % 2 == streamId % 2) {
```
Comment (yschimke) [inline src/StandardReactiveSocket.cpp]: TIOLI: this block inside "streamId != 0" seems involved enough to pull out.

**Comment by yschimke** [comment]
LGTM

File: src/StandardReactiveSocket.cpp (hunk 2 above)
Comment (lehecka) [inline src/StandardReactiveSocket.cpp]: this chunk of code will be moved to a separate class. I already started working on it.

File: src/ConnectionAutomaton.cpp [hunk 3]
```diff
@@ -345,11 +343,11 @@ void ConnectionAutomaton::processFrameImpl(
...
     folly::exception_wrapper ex,
     StreamCompletionSignal signal) {
-  auto thisPtr = this->shared_from_this();
-  auto movedEx = folly::makeMoveWrapper(ex);
-  runInExecutor([thisPtr, movedEx, signal]() mutable {
-    thisPtr->onTerminalImpl(movedEx.move(), signal);
-  });
+  runInExecutor([
+    thisPtr = this->shared_from_this(),
+    movedEx = std::move(ex),
```
Comment (yschimke) [inline src/ConnectionAutomaton.cpp]: nit: do we need the moved prefix here?  What are our conventions for names with lambdas?

**Comment by yschimke** [comment]
Nice!

**Pull request merged**